### PR TITLE
Opening hours fixes

### PR DESCRIPTION
**Pull request by DanielHeath**
Still outstanding
- [ ] Services which are open 24 hours show as "Until 12:00am"
- [ ] No support / testing of public holiday times
- [ ] No 'All times' link on the service pane

### CODE REVIEW

**Comment by danni** [comment]
Tests?

**Comment by DanielHeath** [comment]
Coming. I think this is the first one which needs regular mocha tests.

**Comment by DanielHeath** [comment]
Also I'd like to move some of the logic into the Service class

**Comment by danni** [comment]
Endorse this POV. That will happen as soon as Travis passes. It passes locally if you want to merge the branch into yours.

File: interfaces/mocha.js [hunk 1]
```diff
@@ -0,0 +1,4 @@
+declare function describe(message: string, impl: Function): void;
+declare function xdescribe(message: string, impl: Function): void;
+declare function it(message: string, impl: Function): void;
+declare function xit(message: string, impl: Function): void;
```
Comment (danni) [inline interfaces/mocha.js:4]: Why do we need these and the moment ones now?

Comment (DanielHeath) [inline interfaces/mocha.js:4]: We were previously not using `describe`/`it` in mocha (and now are).

File: src/iss/ServiceOpening.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+        var startOfToday = moment(now()).startOf('day');
+
+        // Turn opening_hours into a sorted list
+        // so that we can scan forwards from the first one
+        // and only find moment()s which are in the future
+        var openingTimes = [];
+        var closingTimes = [];
+        for (var i = 0; i < properties.opening_hours.length; i++) {
+            var oh: issOpeningHours = properties.opening_hours[i];
```
Comment (danni) [inline src/iss/ServiceOpening.js]: Consider `opening_hours`

File: src/iss/ServiceOpening.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        day.format("YYYY-MM-DD") +
+        ' ' + time,
+        'YYYY-MM-DD HH:mm:ss'
+    );
+}
+
+export default class ServiceOpening {
+    now_open: ?boolean;
+    _opening_times: Array<serviceOpeningHours>;
+    _closing_times: Array<serviceOpeningHours>;
```
Comment (danni) [inline src/iss/ServiceOpening.js:41]: Newline here.

File: src/iss/ServiceOpening.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            // split opening / closing hours into separate arrays?
+            if (close.isBefore(moment(now()))) {
+                close.add(1, 'weeks');
+            }
+
+            if (open.isBefore(moment(now()))) {
+                open.add(1, 'weeks');
+            }
+
+            /* flow:disable */
```
Comment (danni) [inline src/iss/ServiceOpening.js]: Why?

File: src/components/OpeningTimes.js [hunk 5]
```diff
@@ -5,98 +5,201 @@ import mui from "material-ui";
...
+            <span className="when">
+                { open.day } { start } { end } ({ open.note })
+            </span>
+        );
+    }
+
+    render(): React.Element {
+        var open = this.props.object.now_open;
+
+        var renderMethod = this.renderUnsure;
```
Comment (danni) [inline src/components/OpeningTimes.js]: Move this into `else`. It makes it clearer.

File: src/iss.js [hunk 6]
```diff
@@ -136,6 +136,16 @@ export class Service {
     _serviceProvisions: Array<string>;
     _siblingServices: searchResults;
 
+
+    /**
+     * open:
+     *
+     * The service opening hours
+     */
+    open(): (ServiceOpening) {
```
Comment (danni) [inline src/iss.js]: Make a property?
Remove extra parenthesis.

File: src/iss/ServiceOpening.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        this.now_open = properties.now_open.now_open;
+
+        var startOfToday = moment(now()).startOf('day');
+
+        // Turn opening_hours into a sorted list
+        // so that we can scan forwards from the first one
+        // and only find moment()s which are in the future
+        var openingTimes = [];
+        var closingTimes = [];
+        for (var i = 0; i < properties.opening_hours.length; i++) {
```
Comment (danni) [inline src/iss/ServiceOpening.js]: `for (var opening_hours of properties.opening_hours)`

File: src/iss/ServiceOpening.js (hunk 2 above)
Comment (DanielHeath) [inline src/iss/ServiceOpening.js]: Would have to be `openingHours` for `jscs`

File: src/iss/ServiceOpening.js [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+            var oh: issOpeningHours = properties.opening_hours[i];
+
+            // Convert (eg) 'Monday' to a Moment()
+            var day = moment(now());
+            day = day.day(oh.day);
```
Comment (danni) [inline src/iss/ServiceOpening.js]: I think you can just do `moment().day(opening_hours.day)`

File: src/iss/ServiceOpening.js [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    "start": Moment,
+}
+
+function timeOfDay(day: Moment, time: string): Moment {
+    return moment(
+        day.format("YYYY-MM-DD") +
+        ' ' + time,
+        'YYYY-MM-DD HH:mm:ss'
+    );
+}
```
Comment (danni) [inline src/iss/ServiceOpening.js:36]: This looks very wrong. Consider `moment.get`/`moment.set`.

File: interfaces/mocha.js (hunk 1 above)
Comment (danni) [inline interfaces/mocha.js:4]: We have been, apparently it's because mocha doesn't have flowtypes?

File: src/iss/ServiceOpening.js (hunk 2 above)
Comment (danni) [inline src/iss/ServiceOpening.js]: Also fine.

File: interfaces/mocha.js (hunk 1 above)
Comment (DanielHeath) [inline interfaces/mocha.js:4]: That's very strange; adding these libs definitely resolved a bunch of flow errors.

File: src/iss/ServiceOpening.js (hunk 3 above)
Comment (DanielHeath) [inline src/iss/ServiceOpening.js:41]: Added.

**Comment by DanielHeath** [comment]
So, this is failing because when we stub the time we don't do it in a timezone-friendly way.

**Comment by DanielHeath** [comment]
@danni ++?

**Comment by danni** [comment]
++

**Pull request merged**